src/components/tl/ucp/tl_ucp_sendrecv.h

```diff
@@ -161,33 +161,52 @@ ucc_tl_ucp_resolve_p2p_by_va(ucc_tl_ucp_team_t *team, void *va, ucp_ep_h *ep,
                              ucc_rank_t peer, uint64_t *rva, ucp_rkey_h *rkey,
                              int *segment)
 {
-    ucc_tl_ucp_context_t *ctx = UCC_TL_UCP_TEAM_CTX(team);
+    ucc_tl_ucp_context_t *ctx            = UCC_TL_UCP_TEAM_CTX(team);
+    ptrdiff_t             key_offset     = 0;
+    const size_t          section_offset = sizeof(uint64_t) * ctx->n_rinfo_segs;
     ucc_rank_t            core_rank;
-    *segment = 0;
+    uint64_t             *rvas;
+    uint64_t             *key_sizes;
+    void                 *keys;
+    ptrdiff_t             base_offset;
+    ucc_context_addr_header_t *h;
+
+    *segment  = -1;
+    core_rank = ucc_ep_map_eval(UCC_TL_TEAM_MAP(team), peer);
+    ucc_assert(UCC_TL_CORE_TEAM(team));
+    peer = ucc_get_ctx_rank(UCC_TL_CORE_TEAM(team), core_rank);
+
+    h = UCC_ADDR_STORAGE_RANK_HEADER(
+        &ctx->super.super.ucc_context->addr_storage, peer);
+    base_offset =
+        (ptrdiff_t)PTR_OFFSET(h, h->components[0].offset + ctx->ucp_addrlen);
```

Comment:
are you assuming that tl/ucp component is loaded first ? (you use directly h->components[0]). I don't think it is correct. You should use a fn: ucc_get_team_ep_addr:

```
 85 static inline void *ucc_get_team_ep_addr(ucc_context_t *context,
  86                                          ucc_team_t *team, ucc_rank_t rank,
  87                                          unsigned long component_id)
  88 {
  89     ucc_context_addr_header_t *h    = ucc_get_team_ep_header(context, team,
  90                                                              rank);
  91     void                      *addr = NULL;
  92     int                        i;
  93     for (i = 0; i < h->n_components; i++) {
  94         if (h->components[i].id == component_id) {
  95             addr = PTR_OFFSET(h, h->components[i].offset);
  96             break;
  97         }
  98     }
  99     ucc_assert(NULL != addr);
 100     return addr;
 101 }
```